Parallel catchup V2

jayz22 opened the pull request:
## What
Resolves https://github.com/stellar/stellar-core-internal/issues/264, by adding:
- `./src/MissionParallelCatchup/parallel_catchup_helm` - a Helm project that encapsulates the core parallel catchup logic -- job queues and core workers -- inside Kubernetes
- `./src/MissionParallelCatchup/job_monitor.py` - an (containerized) application monitoring the component statuses, exposing them via an HTTP endpoint `/status`.
- `MissionHistoryPubnetParallelCatchupV2` - a new mission  which manages the Helm project and monitors job status.
- As pass-by improvement, upgrades `.NET` version to 8.0, and brings `stellar-dotnet-sdk` and `KubernetesClient` to the latest versions.

To try the new mission, run this command : 
`$ dotnet run mission HistoryPubnetParallelCatchupV2 --image=docker-registry.services.stellar-ops.com/dev/stellar-core:latest --pubnet-parallel-catchup-num-workers=6 --pubnet-parallel-catchup-starting-ledger=52500000` 
on your local machine (you need to first have access to an Kubernetes cluster, e.g. via sshuttle to the ssc)

## Why

The current parallel catchup mission frequently suffers from instability (e.g. see https://github.com/stellar/stellar-supercluster/issues/280 and https://github.com/stellar/supercluster/pull/150). The main reason is we are managing the lifecycle of pods, managing the job queue and performing status monitoring all inside the mission itself (outside Kubernetes cluster, by constantly querying Kubernetes API. There are many potential points of failure (e.g. a pod fails, a pod cannot be allocated, the K8s API times out) each would result in a total mission failure. 

By moving all of the core logic inside Kubernetes (leveraging k8s's rich resource and workload management capabilities) this design is more resilient to any single point of failure. It also eliminates virtually all API queries, by moving the status checking query on the job monitor HTTP endpoint (a single HTTP (*not* API) query every 30 secs), thus eliminating a main source of unreliability.

Besides, there are a few other improvements/benefits:
- Resilience and partial progress - The worker set can be partially allocated and make progress (when the cluster is resource restrained). Any pod failure due to Kubernetes will be retried later. 
- Eliminates the overhead of pod creation/deletion, saving the parallel catchup ~3hrs (or ~8%) of total run time.
- Stores worker timing metrics, making them available in Prometheus
- Mostly static (with customizable inputs via `value.yaml`) set of yaml files defining the Kubernetes objects -- better readability and maintainability
- Separates out range generation script, making it easy for future optimization on the range algorithm
- Tweaked resource request/limit of worker pods to improve performance (this is still an ongoing work)

Check out the [Design doc](https://docs.google.com/document/d/1qQpRaRqQWG6JP13voX-cRE2RLCuAJZQz6Wia-lzlMtU/edit?pli=1#heading=h.ggca9nms6hha) for more details

sisuresh commented on src/FSLibrary/MissionHistoryPubnetParallelCatchupV2.fs:
```diff
@@ -0,0 +1,195 @@
+
+System.AppDomain.CurrentDomain.ProcessExit.Add(fun _ -> cleanup ())
+
+Console.CancelKeyPress.Add
```
Supercluster has a [clean](https://github.com/stellar/supercluster/blob/9bd3d8aadd8550158e0baf7b6783f7fb217014ef/src/App/Program.fs#L469-L544) command that we use in [Jenkins](https://github.com/stellar/stellar-supercluster/blob/914c4a867f2e383a7ba3ee2c5f35398d25fa249d/Jenkinsfile#L582) to make sure we teardown properly. Is there any reason to use these custom handlers instead of using that clean path?

sisuresh replied:
I don't think I understand your point - the `clean` command should clean up `StatefulSet` as well right?

jayz22 replied:
The jenkins clean is a separate command that cleans up all the artifacts of the entire pipeline, but not after each step (mission) correct?
Here the cleanup function makes sure the clean up happens after the parallel catchup, otherwise the statefulset resources will be occupied for the missions after it.

sisuresh replied:
Ah that's right. We still do cleanup (I think by calling `StellarFormations` `Cleanup` method after a mission completes. Wouldn't it make sense to put this code in there? We already clean up pods between missions without calling the `clean` command.

jayz22 replied:
I'm not using a `StellarFormation` any more, everything is directly managed under Helm now.

sisuresh reviewed (commented):
Nice work! Just some initial comments. I still need to learn about the new tools being used in this PR.

sisuresh commented on doc/missions.md:
```diff
@@ -52,6 +52,9 @@ Perform full catchup to Public Network, but in parallel to improve performance.
 
 Perform full parallel catchup to Public Network, but with increased parallelism.
 
```
Putting a comment here so we don't forget - getting-started.md needs to be updated to mention that `helm` needs  to be installed for the v2 parallel catchup mission.

jayz22 replied:
Updated the docs to mention `Helm`

jacekn commented on src/MissionParallelCatchup/parallel_catchup_helm/values.yaml:
```diff
@@ -0,0 +1,31 @@
+
+monitor:
+  logging_interval_seconds: 10
+  request_cpu: "100m"
+  request_memory: "100Mi"
```
As above it could be nice to switch to the nested `resources` block

jayz22 replied:
done

jayz22 commented on src/MissionParallelCatchup/parallel_catchup_helm/templates/job_preload_redis.yaml:
```diff
@@ -0,0 +1,57 @@
+    spec:
+      containers:
+      - name: preload
+        image: redis:latest
```
Yeah this is a good point. 
However I just realize different ubuntu distributions supports different versions for redis-cli, for example focal only [supports](https://packages.ubuntu.com/focal/redis-tools) version 5.0.7.
I wonder should we pin it to `redis:5` instead, to make sure redis everywhere (redis-server, core-worker, job-monitor) runs the same version (even for stellar-core from a higher distribution)? 
Or does version discrepancy even matter?

jayz22 replied:
Hmm, I don't think pining the version would work. focal has redis 5 while jammy has redis 6.
Maybe we should just use `redis:7` here so we are taking advantage of newer Redis features and it's recent than both of them. 
In the future we will have to update stellar-core's distribution and we won't have to change this version here frequently.
WDYT?

jayz22 replied:
Changed to `redis:7`

jacekn replied:
We should be fine using jammy base to get redis 6 client. If that works with redis server 7 hat should be find.

jayz22 commented on src/MissionParallelCatchup/parallel_catchup_helm/files/logarithmic_range_generator.sh:
```diff
@@ -0,0 +1,42 @@
+latestLedgerNum=$(echo "$LATEST_LEDGER_NUM" | awk '{printf "%d", $1}')
+numParallelism=$NUM_PARALLELISM
+
+echo "starting logarithmic range generationg with the following inputs: 
```
done

jayz22 commented on src/MissionParallelCatchup/parallel_catchup_helm/files/logarithmic_range_generator.sh:
```diff
@@ -0,0 +1,42 @@
+        echo "${el}/${ledgersToApply}";
+        # our queue assumes push-left-pop-right, but since we are generating the ranges in reverse order, here we push right
+        redis-cli -h redis -p 6379 RPUSH ranges "${el}/${ledgersToApply}";
+        el=$(( el - ledgersToApply ));
```
good catch! this was a mistake, should subtract by `ledgersPerJob` instead (without adding overlapLedgers) as you pointed out, fixed.

jayz22 commented on src/MissionParallelCatchup/parallel_catchup_helm/files/logarithmic_range_generator.sh:
```diff
@@ -0,0 +1,42 @@
+latestLedgerNum=$latestLedgerNum
+numParallelism=$numParallelism"
+
+generate_uniform () {
```
yes this is a routine very similar to `uniform_range_generator.sh`. 
I chose to make them separate so that they can be more cleanly separated and evolve at different pace, the `uniform_range_generator.sh` is identical logic to the current mission's [getCatchupRange](https://github.com/stellar/supercluster/blob/main/src/FSLibrary/MissionCatchupHelpers.fs#L185) and won't be changing much.

jacekn replied:
Should we rename the function in the `logarithmic_range_generator.sh` to `generate_logarithmic`?
Right now we have a script called logarithmic that internally uses the `unifrom` function.

jayz22 replied:
The logarithmic range generator partitions the entire history into logarithmic sized chunks, within each chunk, it equally divides the ranges into number of workers, that's why it's calling generate_uniform here (in the outer scope it divides the remaining ledgers by half before calling generate_uniform). 
In other words the range sizes aren't fully logarithmic, but more like "piecewise logarithmic, locally uniform".

jayz22 commented on src/MissionParallelCatchup/parallel_catchup_helm/templates/catchup_workers.yaml:
```diff
@@ -0,0 +1,92 @@
+    spec:
+      containers:
+      - name: stellar-core
+        image: jayz22/core-worker
```
yes, this is done! 
With the new stellar-core docker image containing redis-cli, any new `-buildtest` package should suffice after https://github.com/stellar/stellar-core/pull/4390 goes in.

jayz22 commented on src/MissionParallelCatchup/parallel_catchup_helm/templates/job_monitor.yaml:
```diff
@@ -0,0 +1,81 @@
+  annotations:
+    nginx.ingress.kubernetes.io/rewrite-target: /
+spec:
+  ingressClassName: private
+  rules:
+  - host: ssc-job-monitor.services.stellar-ops.com
```
hmm I'm not sure about this, I think the host domain is just a static string. Maybe from less code copying perspective yes but I'm not sure if this needs to be configurable.

jacekn replied:
You can turn this in to a parameter set via the value file and have default value there.
So you won't have to set it explicitly but you'll be able to if needed.
But it's not a big deal, we can leave it as is too.

jayz22 replied:
done

jayz22 commented:
I did a round of addressing comments, will continue addressing the rest of them tomorrow. Meanwhile I've started another production workload run, will see if the recent improvements resulted in more timing improvements.

graydon commented on src/MissionParallelCatchup/Dockerfile.jobmonitor:
```diff
@@ -0,0 +1,20 @@
+# Use an official Python runtime as a parent image
+FROM --platform=linux/amd64 python:3.12-slim as build
```
python:3.12-slim appears to be based on debian-slim. I'm not sure if we have a strong debian-vs-ubuntu preference but I suspect everything will work on both. @jacekn do you prefer everything sticks to ubuntu derivatives?

jayz22 replied:
I have opted for ubuntu 22.04, which uses python3:10. As far as I understand the slim version has reduced image size making it better for production, which I don't think matters here.

graydon commented on src/FSLibrary/MissionHistoryPubnetParallelCatchupV2.fs:
```diff
@@ -0,0 +1,195 @@
+        let response =
+            client
+                .GetStringAsync(
+                    "http://ssc-job-monitor.services.stellar-ops.com/status"
```
Is this a stable DNS name? If so how? What if there are 2 SSC jobs running at the same time?

jayz22 replied:
In general we shouldn't have multiple SSC jobs running. If the need comes, I think it's easy to modify the DNS entry to add a wildcard matching to the CNAME record, or to route traffic to different job-monitor services via `paths` .

graydon commented on src/MissionParallelCatchup/parallel_catchup_helm/files/logarithmic_range_generator.sh:
```diff
@@ -0,0 +1,47 @@
+#!/bin/sh
+
+floorSize=$LOGARITHMIC_FLOOR_LEDGERS
+overlapLedgers=$OVERLAP_LEDGERS
+startLedger=$(echo "$STARTING_LEDGER" | awk '{printf "%d", $1}')
+latestLedgerNum=$(echo "$LATEST_LEDGER_NUM" | awk '{printf "%d", $1}')
+numParallelism=$NUM_PARALLELISM
```
I know there's only so much defensive coding one can do in a shell script but it helps to check for variables being defined before use and exiting with an error if not.

jayz22 replied:
done

graydon commented on src/MissionParallelCatchup/parallel_catchup_helm/files/logarithmic_range_generator.sh:
```diff
@@ -0,0 +1,47 @@
+#!/bin/sh
```
If you've got python on these hosts, I would consider writing some or all of these scripts in python. Bourne shell works of course, but even simple things quickly get hard and brittle and difficult to debug.

jayz22 replied:
unfortunately the range generator is based off the Redis image and it doesn't contain Python.

jayz22 replied:
as we try other more complex range strategies, it'd be worth it to move all these different strategies inside a separately Python container.

graydon commented on src/MissionParallelCatchup/parallel_catchup_helm/files/worker.sh:
```diff
@@ -0,0 +1,61 @@
+    exit 1
+    fi
+
+    transaction_sum=$(tac "$LOG_FILE" | grep -m 1 -B 11 "metric 'ledger.transaction.apply':" | grep "sum =" | awk '{print $NF}')
```
I don't understand how this prints the sum of transactions. awk's NF variable is just the number of fields in the input. maybe I'm misunderstanding the point of this line?

graydon replied:
oh NF is the number of fields and $NF is the last argument

graydon commented on src/MissionParallelCatchup/parallel_catchup_helm/templates/catchup_workers.yaml:
```diff
@@ -0,0 +1,89 @@
+  name: worker-script
+data:
+  worker.sh: |-
+    {{- (.Files.Get "files/worker.sh") | nindent 4 }}
```
Oh wow, _indenting the input_ so it becomes a literal in the YAML? Is this really the best that can be done?

graydon reviewed (changes requested):
A few nits noted in comments above, nothing critical.

Looking at the finished product .. I also just remain a little hesitant about this whole thing: the configuration layering is such that you've now got at least 3 layers of nested untyped and not-especially-debuggable languages (The helm templating system, then YAML, then the final bunch of bourne shell / python scripts and sed / awk / redis command / whatever strings).

This is kinda what we were dealing with in the pre-supercluster days (back then it was ruby, some other templating system, and dockerfiles) and one of the motives for consolidating all the logic in supercluster's F# code: there are real types, real variables your IDE and compiler can find and complain about mismatches between, you're not as likely to have to keep re-running the whole job just to find typos and nested-quoting mistakes.

I guess if the F# and API-driven approach is sufficiently alien-seeming, relative to the templates-of-YAML-driven style of Kubernetes configuration that's standard in ops environments, then .. I won't stop you. But I guess I just want to double check with you, @jacekn, @sisuresh, @marta-lokhova and anyone else who's been hacking supercluster for a while: we're all ok continuing on this path, having spent some time on it and looking critically at the result?

(This is not a total objection, just a slightly hesitation; in general I think you did great work here and it's exactly what you said you were going to build and I agreed was an experiment worth conducting! I am just wondering how you and others feel about the result. I'm a bit on the fence. If it seems it might be preferable to several folks?)

sisuresh commented:
> A few nits noted in comments above, nothing critical.
> 
> Looking at the finished product .. I also just remain a little hesitant about this whole thing: the configuration layering is such that you've now got at least 3 layers of nested untyped and not-especially-debuggable languages (The helm templating system, then YAML, then the final bunch of bourne shell / python scripts and sed / awk / redis command / whatever strings).
> 
> This is kinda what we were dealing with in the pre-supercluster days (back then it was ruby, some other templating system, and dockerfiles) and one of the motives for consolidating all the logic in supercluster's F# code: there are real types, real variables your IDE and compiler can find and complain about mismatches between, you're not as likely to have to keep re-running the whole job just to find typos and nested-quoting mistakes.
> 
> I guess if the F# and API-driven approach is sufficiently alien-seeming, relative to the templates-of-YAML-driven style of Kubernetes configuration that's standard in ops environments, then .. I won't stop you. But I guess I just want to double check with you, @jacekn, @sisuresh, @marta-lokhova and anyone else who's been hacking supercluster for a while: we're all ok continuing on this path, having spent some time on it and looking critically at the result?
> 
> (This is not a total objection, just a slight hesitation; in general I think you did great work here and it's exactly what you said you were going to build and I agreed was an experiment worth conducting! I am just wondering how you and others feel about the result. I'm a bit on the fence. If it seems it might be preferable to several folks?)

I think what Jay has here works well for jobs like parallel catchup where we had very little client side logic (that client side logic was mainly just orchestration too). I think F# should still be used for all of our other missions for ease of development and debugging, but if this V2 pubnet parallel catchup improves the reliability of the existing mission, then I think the tradeoff is worth it.

jayz22 commented on src/MissionParallelCatchup/parallel_catchup_helm/files/stellar-core.cfg:
```diff
@@ -0,0 +1,33 @@
+NETWORK_PASSPHRASE="Public Global Stellar Network ; September 2015"
+DATABASE="sqlite3:///data/stellar.db"
+BUCKET_DIR_PATH="/data/buckets"
+LOG_FILE_PATH="/data/stellar-core-{datetime:%Y-%m-%d_%H-%M-%S}.log"    
```
done

jayz22 commented on src/MissionParallelCatchup/parallel_catchup_helm/templates/job_preload_redis.yaml:
```diff
@@ -0,0 +1,57 @@
+           else 
+             echo 'Error: Unknown strategy' && exit 1
+           fi
+        env:
```
done

jayz22 commented on src/MissionParallelCatchup/parallel_catchup_helm/templates/redis_queue.yaml:
```diff
@@ -0,0 +1,37 @@
+    spec:
+      containers:
+      - name: redis
+        image: redis:latest
```
done

jayz22 commented on src/MissionParallelCatchup/parallel_catchup_helm/values.yaml:
```diff
@@ -0,0 +1,31 @@
+  request_ephemeral_storage: "35Gi"
+  limit_cpu: "8"
+  limit_memory: "48Gi"
+  limit_ephemeral_storage: "40Gi"
```
done

jayz22 commented on src/MissionParallelCatchup/parallel_catchup_helm/templates/job_monitor.yaml:
```diff
@@ -0,0 +1,81 @@
+    spec:
+      containers:
+      - name: job-monitor
+        image: jayz22/job-monitor:latest
```
yes, I will parameterize this once the stellar public docker repo has been setup. For now I think it's sufficient for testing.

jayz22 replied:
done

jayz22 commented on src/MissionParallelCatchup/parallel_catchup_helm/templates/catchup_workers.yaml:
```diff
@@ -0,0 +1,92 @@
+      - name: stellar-core
+        image: jayz22/core-worker
+        # resource specs copied from the old supercluster mission
+        resources:
```
I think it's worth specifying the resources (not commented out). Leaving them optional require adding if statements in yaml checking if the field is null, which makes it more gnarly. Also keeps it consistent with the current parallel catchup mission which has always specified resources.

jacekn replied:
Fair enough. We can change it later if needed

jayz22 commented on src/MissionParallelCatchup/job_monitor.py:
```diff
@@ -0,0 +1,117 @@
+
+# Configuration
+REDIS_HOST = 'redis'
+REDIS_PORT = 6379
```
done

jayz22 commented:
I've pushed a few changes that should address all the outstanding comments. 
Besides what's already mentioned in the comments, here are a few main improvements since the last review:
- Simplified parameter overriding in the F# mission. Parameters can be directly overridden via the command line, and it is additive to the `values.yaml` file.
- Improved context in the job monitor messages: for "in-progress" and "failed" queue it now includes the actual ranges (not just the count). Entries in the failed queue are also enhanced with the pod name for better traceability. Worker statuses contain an additional field with the detailed status from `stellar-core:info`.  an example response now looks like:
```
[00:03:11 INF] job monitor response: {"num_remain": 560, "num_succeeded": 0, "jobs_failed": [118400/1920|stellar-core-1 ], "jobs_in_progress": ["52516223/16320", "52548223/16320", "52596223/16320"], "workers": [{"worker_id": 0, "status": "running", "info": ["Catching up to ledger 52596223: Waiting: apply-buckets : 0/1 children completed"]}, {"worker_id": 1, "status": "running", "info": ["Catching up to ledger 52548223: Waiting: apply-buckets : 0/1 children completed"]}, {"worker_id": 2, "status": "running", "info": ["Catching up to ledger 52516223: Waiting: apply-buckets : 0/1 children completed"]}]}
```
- Implemented copying of the log stream of the pod owning a failed range, and dumping into local file.

There were a couple of experiments:
- Tried a full-history catchup with "piecewise logarithmic" partitioning method (see `logarithmic_range_generator.sh`). The result is much, much worse than the current uniform partitioning method. The hope was that ledger apply time roughly follows exponential trend throughout history, thus having more ranges in the front part can reduce the overall weight of bucket download&apply time (making the total time for different ranges more even). In reality the ledger apply time distribution is more complicated, and some middling ranges (with increased range length than before) takes forever to finish, resulting in highly uneven workload distribution. So I've switched back to the normal uniform range generation for now, more experiment (and data) is needed for better optimization. 
- Experimented with writing the core-worker script in Python. This is possible because the stellar-core image already contains Python, so we can directly invoke the script from the command (no need for a new docker image with the Python code). This did not make the script any simpler. The worker script mostly performs unix command (redis-cli, file processing, os calls), moving them into Python made it worse (some basic file processing commands such as `tac` and `ls -t` does not have a standard Python variant and has to be manually implemented), and it's left with mostly a bunch of subprocess calls. I did not commit the python version of the worker. 


> A few nits noted in comments above, nothing critical.
> 
> Looking at the finished product .. I also just remain a little hesitant about this whole thing: the configuration layering is such that you've now got at least 3 layers of nested untyped and not-especially-debuggable languages (The helm templating system, then YAML, then the final bunch of bourne shell / python scripts and sed / awk / redis command / whatever strings).
> 
> This is kinda what we were dealing with in the pre-supercluster days (back then it was ruby, some other templating system, and dockerfiles) and one of the motives for consolidating all the logic in supercluster's F# code: there are real types, real variables your IDE and compiler can find and complain about mismatches between, you're not as likely to have to keep re-running the whole job just to find typos and nested-quoting mistakes.
> 
> I guess if the F# and API-driven approach is sufficiently alien-seeming, relative to the templates-of-YAML-driven style of Kubernetes configuration that's standard in ops environments, then .. I won't stop you. But I guess I just want to double check with you, @jacekn, @sisuresh, @marta-lokhova and anyone else who's been hacking supercluster for a while: we're all ok continuing on this path, having spent some time on it and looking critically at the result?
> 
> (This is not a total objection, just a slight hesitation; in general I think you did great work here and it's exactly what you said you were going to build and I agreed was an experiment worth conducting! I am just wondering how you and others feel about the result. I'm a bit on the fence. If it seems it might be preferable to several folks?)

@graydon to address your comment. In my opinion the benefits of the new approach (I listed out some of them on [top](https://github.com/stellar/supercluster/pull/178#issue-2406517898)) out-weights the cons you mentioned. The parallel catchup mission is a fairly straightforward and static setup, and the Helm/Yaml approach made the structure clear and more visual. All of the container logics involving shell scripts are fairly small and manageable (including the worker logic). 

The extra layer of configuration setting can appear a little intimidating at first. But some of this is because we wanted to have a single point of entry for all supercluster missions. I'd like to point out, there is an extra layer which is the Jenkins job, so the overall pipeline is: 
`Jenkins --> (dotnet) parallel-catchup mission --> Helm config --> pod configs`. 
But we don't really need the dotnet layer (we need it because of single point of entry with other missions), as Jenkins can directly config Helm and launch project, so in the future (if we want) we might spin out the parallel catchup mission to run completely standalone inside Kubernetes (which wouldn't be possible with other missions), Jenkins can directly configure it via the Helm command line and that shall simply the overall flow a lot. 

For other more intricate setups (such as experimenting with topology), F# programmability still offers tremendous benefits and likely can't be replaced.
